ipn/ipnlocal/serve.go

```diff
@@ -37,6 +42,166 @@ type serveHTTPContext struct {
 	DestPort uint16
 }
 
+type serveListener struct {
+	ap     netip.AddrPort
+	ctx    context.Context    // valid while listener is desired
+	cancel context.CancelFunc // for ctx, to close listener
+	logf   logger.Logf
+
+	ln net.Listener     // non-nil if listener is active
+	bo *backoff.Backoff // for retrying failed Listen calls
+}
+
+func newServeListener(ctx context.Context, ap netip.AddrPort, logf logger.Logf) *serveListener {
+	ctx, cancel := context.WithCancel(ctx)
+	return &serveListener{
+		ap:     ap,
+		ctx:    ctx,
+		cancel: cancel,
+		logf:   logf,
+
+		bo: backoff.NewBackoff("serve-listener", logf, 30*time.Second),
+	}
+}
+
+func (s *serveListener) Close() error {
+	s.cancel()
+	return s.ln.Close()
+}
+
+func (s *serveListener) Run(h func(net.Conn)) {
+	for {
+		if s.ctx.Err() != nil {
+			// context cancelled before we could listen
+			return
+		}
+		ln, err := net.Listen("tcp", s.ap.String())
+		if err != nil {
+			s.logf("serve failed to listen on %v, backing off: %v", s.ap, err)
+			s.bo.BackOff(s.ctx, err)
+			continue
+		}
+		s.ln = ln
+		break
+	}
+
+	s.logf("serve listening on %v", s.ap)
+	go s.handleServeListenersAccept(h)
```

Comment:
why a new goroutine for the accepts? Once you have a listener, have a accept loop as long as the listener works, and then if your listener breaks (and context is still good) you can loop back around to doing another listen